File: week_06/jiwon/leet_402.java
import java.util.*;

class Solution {
    public String removeKdigits(String num, int k) {

        Stack<Character> stack = new Stack<>();

        for(int i = 0; i < num.length(); i++) {
            char cur = num.charAt(i);

            // 앞자리에 작은 수 오도록 - 0은 패스
            while(!stack.isEmpty() && k > 0 && stack.peek() > cur && stack.peek() != '0') {
                stack.pop();
                k--;
            }

            stack.push(cur);
        }
        
        while(k > 0) {
            stack.pop();
            k--;
        }

        StringBuilder sb = new StringBuilder();

        for (char c : stack) {
            sb.append(c);
        }

        // 앞쪽에 채워진 0 지워서 출력
        int idx = 0;

        while(idx < sb.length() && sb.charAt(idx) == '0') {
            idx++;
        }

        // 비어있는 경우 "0" 출력
        if(sb.length() == idx) {
            return "0";
        }
        
        return sb.substring(idx);
    }
}

File: week_06/jaehan/LEET_402.py
## k 숫자 제거
## 그리디 알고리즘

class Solution(object):
    def removeKdigits(self, num, k):
        
        stack = []
        
        for n in num:
            
            while stack and k > 0 and stack[-1] > n :
                stack.pop()
                k -= 1
                
                
            stack.append(n)
            
        # 아직 제거할 숫자가 존재할 경우 뒤에서 제거
        if k > 0 :
            stack = stack[:-k]
        
        # 리스트 문자열로 합치기
        # 맨 앞 "0" 제거
        # 만약 결과가 빈 문자열이면 0 리턴
        result = ''.join(stack).lstrip("0")
        result = result if result else "0"
            
        # print(result)
        return result
    
    
sol = Solution()
sol.removeKdigits("10200",1)
            
        